File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        TreeNode *temp = root;
        vector<int> res;

        if (root == nullptr) {
            return res;
        }

        vector<vector<int>> tree; 
        queue<pair<TreeNode*, int>> bfs;

        bfs.push(make_pair(root, 0));
        while (!bfs.empty()) {
            pair<TreeNode*, int> curr = bfs.front();
            bfs.pop();
            if (curr.first->left != nullptr) {
                bfs.push(make_pair(curr.first->left, curr.second + 1));
            }
            if (curr.first->right != nullptr) {
                bfs.push(make_pair(curr.first->right, curr.second + 1));
            }
            if (tree.size() < curr.second + 1) {
                vector<int> temp;
                tree.push_back(temp);
            }
            tree[curr.second].push_back(curr.first->val);
        }

        for (int i = 0; i < tree.size(); i++) {
            vector<int> temp = tree[i];
            res.push_back(*(temp.end() - 1));
        }
        return res;
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        stack = [root]
        mp = {None: (False, False)}

        while stack:
            node = stack[-1]

            if node.left and node.left not in mp:
                stack.append(node.left)
            elif node.right and node.right not in mp:
                stack.append(node.right)
            else:
                stack.pop()
                
                containsP = mp[node.left][0] or mp[node.right][0]
                containsQ = mp[node.left][1] or mp[node.right][1]

                if node.val == p.val:
                    containsP = True
                if node.val == q.val:
                    containsQ = True
                if containsP and containsQ:
                    return node
                mp[node] = (containsP, containsQ)
        return root